### PR TITLE
Issue diagnostic warnings when context types don't indicate serializable types

### LINKED ISSUE #58698: System.Text.Json code gen: unhelpful error when custom serializer context is not decorated with any JsonSerializable
When my custom serialiazer context class has zero JsonSerializable attributes, I get an error message which does not help me understand what the problem is.

I simply get the dreaded:
```
CS0534	'MyJsonSerializerContext' does not implement inherited abstract member 'JsonSerializerContext.GetTypeInfo(Type)'
```

I do not get any additional information to help me realize that I forgot to add JsonSerializable attributes to my custom class. It would be extremely helpful to get an error message similar to `Your custom serializer context class must have at least one JsonSerializable attribute`.

Repro:

```csharp
public static class Program
{
    public static void Main()
    {
        Console.WriteLine("Hello World!");
    }
}

// Notice there are no JsonSerializable attributes here.
internal partial class MyJsonSerializerContext : JsonSerializerContext
{
}
```

As an aside: Is there a way to get more detailed information when something goes wrong with the code generator? Does it generate a log somewhere?

### PR DESCRIPTION
**Pull request by layomia**
Fixes https://github.com/dotnet/runtime/issues/58698. I propose porting to 6.0 for improved user experience for the src-gen feature.

With this PR we'll start issuing a diagnostic warning when a context type doesn't indicate serializable types. This warning is only logged when the context type contains no members. We assume that the user wants source generation to provide an implementation, e.g.

```
internal partial class MyJsonSerializerContext : JsonSerializerContext
{
}
```

In other cases (i.e. if the context type has at least one member), the user might have a handwritten context for which they don't need source generation. In those cases we don't issue a warning, e.g this example from our [tests](https://github.com/dotnet/runtime/blob/2f40a52dc742739f60665f0604ba24ee25cf4000/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/ContextClasses.cs#L49):

```cs
internal partial class JsonContext : JsonSerializerContext
{
    public override JsonTypeInfo? GetTypeInfo(Type type)
    {
        if (type == typeof(JsonMessage))
        {
            return JsonMessage;
        }

        return null;
    }

    public JsonTypeInfo<JsonMessage> JsonMessage => ...;
}   
```

### CODE REVIEW

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs [hunk 1]
```diff
@@ -407,31 +423,25 @@ private static bool TryGetClassDeclarationList(INamedTypeSymbol typeSymbol, [Not
                 return typeGenerationSpec;
             }
 
-            internal static bool IsSyntaxTargetForGeneration(SyntaxNode node) => node is ClassDeclarationSyntax { AttributeLists: { Count: > 0 }, BaseList: { Types : {Count : > 0 } } };
+            internal static bool IsSyntaxTargetForGeneration(SyntaxNode node) => node is ClassDeclarationSyntax { BaseList: { Types : { Count : > 0 } } };
```
Comment (layomia) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:426]: For diffs between here and EOF, I'm relaxing the criteria for types we consider for generation

**Before**: include all classes that are annotated with a type called "JsonSerializableAttribute", and also derive or implement at least one type.
**After**: include all classes that derive from a type called "JsonSerializerContext".

I believe this change is fine since we'll have about the same number of types to consider in most source gen scenarios.

Comment (eerhardt) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:426]: Note that this could impact perf. I would test the before/after of this change and ensure for large solutions that target `net6.0` you aren't negatively impacting the perf.

I can get you a large solution offline, if you need it to test.

File: src/libraries/System.Text.Json/gen/Resources/Strings.resx [hunk 2]
```diff
@@ -130,7 +130,7 @@
     <value>Did not generate serialization metadata for type.</value>
   </data>
   <data name="ContextClassesMustBePartialMessageFormat" xml:space="preserve">
-    <value>Derived 'JsonSerializerContext' type '{0}' specifies JSON-serializable types. The type and all containing types must be made partial to kick off source generation.</value>
+    <value>Derived 'JsonSerializerContext' type '{0}' and all containing types must be made partial to kick off source generation.</value>
```
Comment (steveharter) [inline src/libraries/System.Text.Json/gen/Resources/Strings.resx:133]: Suggest rewording to avoid "kick off source generation":
"..and all containing types must be made partial for source generation."

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs [hunk 3]
```diff
@@ -193,19 +201,27 @@ public Parser(Compilation compilation, in SourceProductionContext sourceProducti
...
+                    INamedTypeSymbol contextTypeSymbol = (INamedTypeSymbol)compilationSemanticModel.GetDeclaredSymbol(classDeclarationSyntax);
+                    Debug.Assert(contextTypeSymbol != null);
+                    string[] diagnosticMessageArgs = new string[] { contextTypeSymbol.Name };
+
                     if (serializableAttributeList == null)
                     {
                         // No types were indicated with [JsonSerializable]
+
+                        // If there are no members on the context, assume that the user wanted source generation and notify them that there are no serializable types.
+                        if (!contextTypeSymbol.MemberNames.Any())
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:213]: My feeling is that we should refine this to check for the existence of `GetTypeInfo` overrides. A user that unintentionally adds _any_ member to their context class (be it a helper method, type, etc.) would see this diagnostic disappear completely.

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs (hunk 1 above)
Comment (steveharter) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:426]: It is possible this change could also improve perf, depending on the individual performance attributes vs. base classes, and of course the actual solution and its types (do they have base classes, attributes, etc).

@eerhardt how do you suggest measuring perf? Manual measuring the VS experience?

Comment (eerhardt) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:426]: @ericstj made a micro benchmark for source generators: https://github.com/ericstj/sample-code/tree/generatorBenchmark

But honestly, I would measure it in VS with a large solution. The way I measured the time taken was to add an EventSource around the source generator implementation. And then use PerfView to see how much time was spent in the source generator when it wasn't doing any work.

But if you are using the latest Roslyn builds (as of 5 days ago), this timing measurement is built in. See https://github.com/dotnet/roslyn/pull/55892.

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:426]: One possible way to mitigate would be check the member count of the type. In pseudocode:
```csharp
internal static bool IsSyntaxTargetForGeneration(SyntaxNode node) =>
    node is ClassDeclarationSyntax { 
            { BaseList: { Types : { Count : > 0 } } and 
            { AttributeLists: { Count: > 0 } } or 
            { MemberList: { Count: 0 } } };
```
Note that this would contradict my feedback in https://github.com/dotnet/runtime/pull/58768/files#r704458155.

File: src/libraries/System.Text.Json/gen/Resources/xlf/Strings.cs.xlf [hunk 4]
```diff
@@ -3,15 +3,25 @@
   <file datatype="xml" source-language="en" target-language="cs" original="../Strings.resx">
     <body>
       <trans-unit id="ContextClassesMustBePartialMessageFormat">
-        <source>Derived 'JsonSerializerContext' type '{0}' specifies JSON-serializable types. The type and all containing types must be made partial to kick off source generation.</source>
-        <target state="translated">Odvozený typ JsonSerializerContext {0} určuje typy serializovatelné na JSON. Typ a všechny obsahující typy musí být částečné, aby se zahájilo generování zdroje.</target>
+        <source>Derived 'JsonSerializerContext' type '{0}' and all containing types must be made partial to kick off source generation.</source>
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/gen/Resources/xlf/Strings.cs.xlf:6]: If we do backport this to `release/6.0` now, isn't there a risk of translations going out of sync?

**Comment by eiriktsarpalis** [comment]
This is a good quality of life improvement, although I'm not necessarily convinced it meets the bar for a 6.0 backport.

File: src/libraries/System.Text.Json/gen/Resources/xlf/Strings.cs.xlf (hunk 4 above)
Comment (ericstj) [reply src/libraries/System.Text.Json/gen/Resources/xlf/Strings.cs.xlf:6]: We have a 1 week SLA for translations in `main`.  We have enough time to backport to `6.0`.  That backport will need to be tracked with an issue so we don't forget.

**Comment by layomia** [comment]
> although I'm not necessarily convinced it meets the bar for a 6.0 backport.

I can see that, but would we consider it a breaking change to go from no diagnostic to issuing one in a future release?

**Comment by ericstj** [comment]
It fails with an error today, right?  I don't think adding a diagnostic to something that already fails to compile is breaking.  Even cases where we add diagnostics to passing code is acceptable in a new release (this is what the roslyn-analyzers repo does with every new analyzer).

Is the case this diagnostic fires always an error case?  In that sense I'm not sure anyone would even see the diagnostic as users tend to prioritize errors over warnings.  Or are we filling in the abstracts with empty implementations?

**Comment by layomia** [comment]
@ericstj, yes this is already a failure case (due to CS0534) so adding a diagnostic in the future wouldn't be breaking.

> Is the case this diagnostic fires always an error case?

It's not always an error case (depending on user intent), and we don't touch the user context unless they indicate serializable types and derive from `JsonSerializerContext`.

I'll close this PR for 6.0 and we can address in 7.
